Q2BSP + QBSP

Paril opened the pull request:
The framework for Q2BSP extended (QBSP) from Qbism's tools. Since it's basically a copy of BSP2, it uses a lot of the functions from there.

The hardest part was needing to pass ident around, because (like Q1BSP) the ident is what is changed here, not the version # which is still 38.

ericwa commented:
Nice!

Looks like `light`'s build is failing, its 'ConvertBSPFormat' calls will need updating too.

Would it be worth using the suffix `_qbism` instead of `_qbsp` for these structs? Might be less confusing in this codebase, although maybe inconsistent with the naming in other tools?

Paril commented:
Yeah qbism would make more sense. I forgot about light since I had to
disable it compiling, Embree has trouble with vcpkg on Windows, it doesn't
want to link for some reason.

On Sat., Aug. 21, 2021, 11:13 a.m. Eric Wasylishen, <
***@***.***> wrote:

> Nice!
>
> Looks like light's build is failing, its 'ConvertBSPFormat' calls will
> need updating too.
>
> Would it be worth using the suffix _qbism instead of _qbsp for these
> structs? Might be less confusing in this codebase, although maybe
> inconsistent with the naming in other tools?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ericwa/ericw-tools/pull/313#issuecomment-903131270>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJLAHU6XXCCBXDQPQGYCGLT567A5ANCNFSM5CRU4BYQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

Paril commented:
Okay, I fixed light and renamed all of the qbsp stuff to qbism.
I think the next thing to do would be to finally fix the loadversion/loadident/version/ident stuff, so we can get rid of that `99` tag and properly convey version info. It won't matter much to Q1, but Q2 (and eventually Q3) will be a lot nicer to work with down the road.

I was also curious, since this sourcebase is like semi-cpp, how far did you plan on taking the C++ness? What's the highest language level I should be allowed to use? I was thinking, regarding conversion & swapping, that the codebase would be a ton simpler if those were stored in the struct itself as conversion operators and swap functions rather than xxxtoyyy_swap/copy functions.

ericwa commented:
That sounds good!
Feel free to use C++ wherever it feels useful. I know it's kind of a mishmash currently. The compiler target is VS2019, gcc/clang versions with C++17 support.

Paril commented:
Okay, cool. I'm currently on C++20 but I'll downgrade to C++17 just to be safe.
I still have no idea how to get Embree to link, though. Any suggestions? I basically just tossed it into my VS, and it did pick up on my vcpkg installation and do all that jazz, but it fails to link:

```
  [57/86] Linking CXX executable light\light.exe
  FAILED: light/light.exe 
  cmd.exe /C "cd . && "A:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=light\CMakeFiles\light.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- "A:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30423\bin\Hostx64\x64\link.exe" /nologo light\CMakeFiles\light.dir\trace_embree.cc.obj light\CMakeFiles\light.dir\entities.cc.obj light\CMakeFiles\light.dir\litfile.cc.obj light\CMakeFiles\light.dir\ltface.cc.obj light\CMakeFiles\light.dir\trace.cc.obj light\CMakeFiles\light.dir\light.cc.obj light\CMakeFiles\light.dir\phong.cc.obj light\CMakeFiles\light.dir\bounce.cc.obj light\CMakeFiles\light.dir\surflight.cc.obj light\CMakeFiles\light.dir\settings.cc.obj light\CMakeFiles\light.dir\imglib.cc.obj light\CMakeFiles\light.dir\__\common\bspfile.cc.obj light\CMakeFiles\light.dir\__\common\entdata.cc.obj light\CMakeFiles\light.dir\__\common\cmdlib.cc.obj light\CMakeFiles\light.dir\__\common\mathlib.cc.obj light\CMakeFiles\light.dir\__\common\qvec.cc.obj light\CMakeFiles\light.dir\__\common\mesh.cc.obj light\CMakeFiles\light.dir\__\common\log.cc.obj light\CMakeFiles\light.dir\__\common\threads.cc.obj light\CMakeFiles\light.dir\__\common\polylib.cc.obj light\CMakeFiles\light.dir\__\common\bsputils.cc.obj light\CMakeFiles\light.dir\main.cc.obj  /out:light\light.exe /implib:light\light.lib /pdb:light\light.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D A:\ericw-tools\out\build\x64-Debug\light && C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -noprofile -executionpolicy Bypass -file D:/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary A:/ericw-tools/out/build/x64-Debug/light/light.exe -installedDir D:/vcpkg/installed/x64-windows/debug/bin -OutVariable out && cd /D A:\ericw-tools\out\build\x64-Debug\light && "A:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E copy_if_different D:/vcpkg/installed/x64-windows/share/embree/doc/LICENSE.txt A:/ericw-tools/out/build/x64-Debug/light/LICENSE-embree.txt""
  LINK Pass 1: command "A:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30423\bin\Hostx64\x64\link.exe /nologo light\CMakeFiles\light.dir\trace_embree.cc.obj light\CMakeFiles\light.dir\entities.cc.obj light\CMakeFiles\light.dir\litfile.cc.obj light\CMakeFiles\light.dir\ltface.cc.obj light\CMakeFiles\light.dir\trace.cc.obj light\CMakeFiles\light.dir\light.cc.obj light\CMakeFiles\light.dir\phong.cc.obj light\CMakeFiles\light.dir\bounce.cc.obj light\CMakeFiles\light.dir\surflight.cc.obj light\CMakeFiles\light.dir\settings.cc.obj light\CMakeFiles\light.dir\imglib.cc.obj light\CMakeFiles\light.dir\__\common\bspfile.cc.obj light\CMakeFiles\light.dir\__\common\entdata.cc.obj light\CMakeFiles\light.dir\__\common\cmdlib.cc.obj light\CMakeFiles\light.dir\__\common\mathlib.cc.obj light\CMakeFiles\light.dir\__\common\qvec.cc.obj light\CMakeFiles\light.dir\__\common\mesh.cc.obj light\CMakeFiles\light.dir\__\common\log.cc.obj light\CMakeFiles\light.dir\__\common\threads.cc.obj light\CMakeFiles\light.dir\__\common\polylib.cc.obj light\CMakeFiles\light.dir\__\common\bsputils.cc.obj light\CMakeFiles\light.dir\main.cc.obj /out:light\light.exe /implib:light\light.lib /pdb:light\light.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:light\CMakeFiles\light.dir/intermediate.manifest light\CMakeFiles\light.dir/manifest.res" failed (exit code 1120) with the following output:
A:\ericw-tools\out\build\x64-Debug\trace_embree.cc.obj : error LNK2019: unresolved external symbol __imp_rtcNewDevice referenced in function "void __cdecl Embree_TraceInit(struct mbsp_t const *)" (?Embree_TraceInit@@YAXPEBUmbsp_t@@@Z)
A:\ericw-tools\out\build\x64-Debug\trace_embree.cc.obj : error LNK2019: unresolved external symbol __imp_rtcGetDeviceProperty referenced in function "void __cdecl Embree_TraceInit(struct mbsp_t const *)" (?Embree_TraceInit@@YAXPEBUmbsp_t@@@Z)
```
etc etc.

TBB seems to be okay, however. I have zero experience with CMake projects in VS and how it integrates with vcpkg
